pinot-common/src/main/java/com/linkedin/pinot/common/utils/helix/HelixHelper.java

```diff
@@ -70,67 +70,67 @@
+          IdealState updatedIdealState;
 
           try {
-            success =
-                baseDataAccessor.set(path, updatedIdealState.getRecord(), idealState.getRecord().getVersion(),
-                    AccessOption.PERSISTENT);
+            updatedIdealState = updater.apply(idealStateCopy);
           } catch (Exception e) {
-            boolean idealStateIsCompressed = updatedIdealState.getRecord().getBooleanField("enableCompression", false);
-
-            LOGGER.warn("Caught exception while updating ideal state for resource {} (compressed={}), retrying.",
-                resourceName, idealStateIsCompressed, e);
+            LOGGER.error("Caught exception while updating ideal state", e);
             return false;
           }
 
-          if (success) {
-            return true;
+          // If there are changes to apply, apply them
+          if (!EqualityUtils.isEqual(idealState, updatedIdealState) && updatedIdealState != null) {
+            BaseDataAccessor<ZNRecord> baseDataAccessor = dataAccessor.getBaseDataAccessor();
+            boolean success;
+
+            // If the ideal state is large enough, enable compression
+            if (MAX_PARTITION_COUNT_IN_UNCOMPRESSED_IDEAL_STATE < updatedIdealState.getPartitionSet().size()) {
+              updatedIdealState.getRecord().setBooleanField("enableCompression", true);
+            }
+
+            try {
+              success =
+                  baseDataAccessor.set(path, updatedIdealState.getRecord(), idealState.getRecord().getVersion(),
+                      AccessOption.PERSISTENT);
+            } catch (Exception e) {
+              boolean idealStateIsCompressed = updatedIdealState.getRecord().getBooleanField("enableCompression", false);
+
+              LOGGER.warn("Caught exception while updating ideal state for resource {} (compressed={}), retrying.",
+                  resourceName, idealStateIsCompressed, e);
+              return false;
+            }
+
+            if (success) {
+              return true;
+            } else {
+              LOGGER.warn("Failed to update ideal state for resource {}, retrying.", resourceName);
+              return false;
+            }
           } else {
-            LOGGER.warn("Failed to update ideal state for resource {}, retrying.", resourceName);
-            return false;
+            LOGGER.warn("Idempotent or null ideal state update for resource {}, skipping update.", resourceName);
+            return true;
           }
-        } else {
-          LOGGER.warn("Idempotent or null ideal state update for resource {}, skipping update.", resourceName);
-          return true;
         }
-      }
-    });
-
-    if (!successful) {
-      throw new RuntimeException("Failed to update ideal state for resource " + resourceName);
+      });
+    } catch (Exception e) {
+      throw new RuntimeException("Caught exception while updating ideal state for resource: " + resourceName, e);
```

Comment:
Correct me if I am wrong, but the earlier piece of code logs the actual exception encountered during attempts, whereas this one will only throw AttemptsExceededException ?

Reply:
This one will log the actual exception encountered during attempts, or AttemptsExceededException only if no exception encountered but run out of attempts